pinot-broker/src/main/java/com/linkedin/pinot/broker/queryquota/TableQueryQuotaManager.java

```diff
@@ -0,0 +1,138 @@
+   * */
+  public void initTableQueryQuota(TableConfig tableConfig, ExternalView brokerResource) {
+    String tableName = tableConfig.getTableName();
+    String rawTableName = TableNameBuilder.extractRawTableName(tableName);
+    LOGGER.info("Initializing rate limiter for table {}", rawTableName);
+
+    CommonConstants.Helix.TableType tableType = TableNameBuilder.getTableTypeFromTableName(tableName);
+    if (tableType == CommonConstants.Helix.TableType.OFFLINE) {
+      QuotaConfig quotaConfig = tableConfig.getQuotaConfig();
+      if (quotaConfig != null && quotaConfig.getQps() != null) {
+        createRateLimiter(tableName, rawTableName, brokerResource, quotaConfig);
+      } else {
+        String realtimeTableName = TableNameBuilder.REALTIME.tableNameWithType(rawTableName);
+        TableConfig realTimeTableConfig = ZKMetadataProvider.getTableConfig(_propertyStore, realtimeTableName);
+        if (realTimeTableConfig == null) {
+          return;
+        }
+        QuotaConfig realTimeTableQuotaConfig = realTimeTableConfig.getQuotaConfig();
+        if (realTimeTableQuotaConfig != null && realTimeTableQuotaConfig.getQps() != null) {
+          createRateLimiter(realtimeTableName, rawTableName, brokerResource, realTimeTableQuotaConfig);
+        }
+      }
+    } else {
+      QuotaConfig quotaConfig = tableConfig.getQuotaConfig();
+      if (quotaConfig != null && quotaConfig.getQps() != null) {
+        createRateLimiter(tableName, rawTableName, brokerResource, quotaConfig);
+      } else {
+        String offlineTableName = TableNameBuilder.OFFLINE.tableNameWithType(rawTableName);
+        TableConfig offlineTableConfig = ZKMetadataProvider.getTableConfig(_propertyStore, offlineTableName);
+        if (offlineTableConfig == null) {
+          return;
+        }
+        QuotaConfig offlineTableQuotaConfig = offlineTableConfig.getQuotaConfig();
+        if (offlineTableQuotaConfig != null && offlineTableQuotaConfig.getQps() != null) {
+          createRateLimiter(offlineTableName, rawTableName, brokerResource, offlineTableQuotaConfig);
+        }
+      }
+    }
+  }
+
+  /**
+   * Create a rate limiter for a table
+   * @param tableName table name with table type.
+   * @param rawTableName raw table name.
+   * @param brokerResource broker resource which stores all the broker states of each table.
+   * @param quotaConfig quota config of the table.
+   * */
+  private void createRateLimiter(String tableName, String rawTableName, ExternalView brokerResource, @Nonnull QuotaConfig quotaConfig) {
+    if (brokerResource == null) {
+      return;
+    }
+    Map<String, String> stateMap = brokerResource.getStateMap(tableName);
+    int onlineCount = 0;
+    for (String state : stateMap.values()) {
+      if (state.equals(CommonConstants.Helix.StateModel.SegmentOnlineOfflineStateModel.ONLINE)) {
+        onlineCount++;
+      }
+    }
+    // Get the dynamic rate
+    double rate = Double.parseDouble(quotaConfig.getQps()) / onlineCount;
```

Comment:
Does `NumberFormatException` get caught anywhere? If not, it will prevent a table with an incorrect table configuration from being served.

Also, what happens if there are no servers online? Or if the table is not in the external view?

Reply:
Good point! I'll adjust the code and add unit tests for it. 👍